ocaml/asmcomp/cmm_helpers.ml

```diff
@@ -80,17 +80,20 @@ let caml_int64_ops = "caml_int64_ops"
 let pos_arity_in_closinfo = 8 * size_addr - 8
        (* arity = the top 8 bits of the closinfo word *)
 
-let closure_info ~arity ~startenv =
+let closure_info ~arity ~startenv ~is_last =
   let arity =
     match arity with
     | Lambda.Tupled, n -> -n
     | Lambda.Curried _, n -> n
   in
   assert (-128 <= arity && arity <= 127);
-  assert (0 <= startenv && startenv < 1 lsl (pos_arity_in_closinfo - 1));
+  assert (0 <= startenv && startenv < 1 lsl (pos_arity_in_closinfo - 2));
   Nativeint.(add (shift_left (of_int arity) pos_arity_in_closinfo)
-                 (add (shift_left (of_int startenv) 1)
-                      1n))
+    (add
+      (shift_left
+          (Bool.to_int is_last |> Nativeint.of_int)
+          (pos_arity_in_closinfo - 1))
+        (add (shift_left (of_int startenv) 1) 1n)))
```

Comment:
Maybe reformat, or add let bindings to intermediate subexpressions (with a name).
This expression is not really easy to follow what is happening